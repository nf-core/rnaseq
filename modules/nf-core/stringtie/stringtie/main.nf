process STRINGTIE_STRINGTIE {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/stringtie:2.2.1--hecb563c_2' :
        'biocontainers/stringtie:2.2.1--hecb563c_2' }"

    input:
    tuple val(meta), path(bam)
    path  annotation_gtf

    output:
    tuple val(meta), path("*.transcripts.gtf"), emit: transcript_gtf
    tuple val(meta), path("*.abundance.txt")  , emit: abundance
    tuple val(meta), path("*.coverage.gtf")   , optional: true, emit: coverage_gtf
    tuple val(meta), path("*.ballgown")       , optional: true, emit: ballgown
    path  "versions.yml"                      , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args      = task.ext.args ?: ''
    def prefix    = task.ext.prefix ?: "${meta.id}"
    def reference = annotation_gtf ? "-G $annotation_gtf" : ""
    def ballgown  = annotation_gtf ? "-b ${prefix}.ballgown" : ""
    def coverage  = annotation_gtf ? "-C ${prefix}.coverage.gtf" : ""

    def strandedness = ''
    if (meta.strandedness == 'forward') {
        strandedness = '--fr'
    } else if (meta.strandedness == 'reverse') {
        strandedness = '--rf'
    }
    """
    stringtie \\
        $bam \\
        $strandedness \\
        $reference \\
        -o ${prefix}.transcripts.gtf \\
        -A ${prefix}.gene.abundance.txt \\
        $coverage \\
        $ballgown \\
        -p $task.cpus \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        stringtie: \$(stringtie --version 2>&1)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.transcripts.gtf
    touch ${prefix}.gene.abundance.txt
    touch ${prefix}.coverage.gtf
    touch ${prefix}.ballgown

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        stringtie: \$(stringtie --version 2>&1)
    END_VERSIONS
    """
}
