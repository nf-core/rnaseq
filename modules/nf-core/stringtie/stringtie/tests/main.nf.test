nextflow_process {

    name "Test Process STRINGTIE_STRINGTIE"
    script "../main.nf"
    process "STRINGTIE_STRINGTIE"
    tag "modules"
    tag "modules_nfcore"
    tag "stringtie"
    tag "stringtie/stringtie"

    test("sarscov2 [bam] - forward strandedness") {

        when {
            process {
                """
                input[0] = [
                            [ id:'test', strandedness:'forward' ], // meta map
                            [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam", checkIfExists: true) ]
                           ]
                input[1] = []
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(process.out.transcript_gtf).match("fs_transcript_gtf") },
            { assert snapshot(process.out.abundance).match("fs_abundance") },
            { assert snapshot(process.out.versions).match("fs_versions") }
            )
        }
    }

    test("sarscov2 [bam] - forward strandedness + reference annotation") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', strandedness:'forward' ], // meta map
                    [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam", checkIfExists: true) ]
                ]
                input[1] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(process.out.transcript_gtf).match("fs_gtf_transcript_gtf") },
            { assert snapshot(process.out.abundance).match("fs_gtf_abundance") },
            { assert snapshot(process.out.ballgown).match("fs_gtf_ballgown") },
            { assert snapshot(process.out.versions).match("fs_gtf_versions") }
            )
        }
    }

    test("sarscov2 [bam] - reverse strandedness") {

        when {
            process {
                """
                input[0] = [
                            [ id:'test', strandedness:'reverse' ], // meta map
                            [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam", checkIfExists: true) ]
                           ]
                input[1] = []
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(process.out.transcript_gtf).match("rs_transcript_gtf") },
            { assert snapshot(process.out.abundance).match("rs_abundance") },
            { assert snapshot(process.out.versions).match("rs_versions") }
            )
        }
    }

    test("sarscov2 [bam] - reverse strandedness + reference annotation") {

        when {
            process {
                """
                input[0] = [
                            [ id:'test', strandedness:'reverse' ], // meta map
                            [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam", checkIfExists: true) ]
                           ]
                input[1] = file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
            { assert process.success },
            { assert snapshot(process.out.transcript_gtf).match("rs_gtf_transcript_gtf") },
            { assert snapshot(process.out.abundance).match("rs_gtf_abundance") },
            { assert snapshot(process.out.ballgown).match("rs_gtf_ballgown") },
            { assert snapshot(process.out.versions).match("rs_gtf_versions") }
            )
        }
    }
}
