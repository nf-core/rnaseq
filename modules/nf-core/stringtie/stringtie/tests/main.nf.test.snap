{
    "fs_abundance": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    "test.gene.abundance.txt:md5,d6f5c8cadb8458f1df0427cf790246e3"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:55:41.032044613"
    },
    "fs_transcript_gtf": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    "test.transcripts.gtf:md5,569137af5be452413086b50653a97203"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:55:41.017978904"
    },
    "rs_abundance": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "reverse"
                    },
                    "test.gene.abundance.txt:md5,d6f5c8cadb8458f1df0427cf790246e3"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:13.601112933"
    },
    "fs_gtf_versions": {
        "content": [
            [
                "versions.yml:md5,3410e8ac349d18c85ddee89337851d38"
            ]
        ],
        "timestamp": "2023-11-23T13:56:00.523797974"
    },
    "fs_gtf_transcript_gtf": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    "test.transcripts.gtf:md5,f56cf8aba2c4a5673bc7963ba5f12d04"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:00.475164879"
    },
    "rs_versions": {
        "content": [
            [
                "versions.yml:md5,3410e8ac349d18c85ddee89337851d38"
            ]
        ],
        "timestamp": "2023-11-23T13:56:13.623892691"
    },
    "rs_gtf_transcript_gtf": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "reverse"
                    },
                    "test.transcripts.gtf:md5,bb346053a8c156b803b055133376c7fa"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:22.693599559"
    },
    "fs_gtf_abundance": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    "test.gene.abundance.txt:md5,7d8bce7f2a922e367cedccae7267c22e"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:00.482135418"
    },
    "rs_gtf_ballgown": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "reverse"
                    },
                    [
                        "e2t.ctab:md5,e981c0038295ae54b63cedb1083f1540",
                        "e_data.ctab:md5,879b6696029d19c4737b562e9d149218",
                        "i2t.ctab:md5,8a117c8aa4334b4c2d4711932b006fb4",
                        "i_data.ctab:md5,be3abe09740603213f83d50dcf81427f",
                        "t_data.ctab:md5,3b66c065da73ae0dd41cc332eff6a818"
                    ]
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:22.715698347"
    },
    "rs_transcript_gtf": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "reverse"
                    },
                    "test.transcripts.gtf:md5,31c34aec2bf36bb0ea3c16c2afeeeb1f"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:13.590054035"
    },
    "rs_gtf_versions": {
        "content": [
            [
                "versions.yml:md5,3410e8ac349d18c85ddee89337851d38"
            ]
        ],
        "timestamp": "2023-11-23T13:56:22.725513476"
    },
    "fs_gtf_ballgown": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "forward"
                    },
                    [
                        "e2t.ctab:md5,e981c0038295ae54b63cedb1083f1540",
                        "e_data.ctab:md5,6b4cf69bc03f3f69890f972a0e8b7471",
                        "i2t.ctab:md5,8a117c8aa4334b4c2d4711932b006fb4",
                        "i_data.ctab:md5,be3abe09740603213f83d50dcf81427f",
                        "t_data.ctab:md5,3b66c065da73ae0dd41cc332eff6a818"
                    ]
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:00.494299817"
    },
    "fs_versions": {
        "content": [
            [
                "versions.yml:md5,3410e8ac349d18c85ddee89337851d38"
            ]
        ],
        "timestamp": "2023-11-23T13:55:41.049417582"
    },
    "rs_gtf_abundance": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "strandedness": "reverse"
                    },
                    "test.gene.abundance.txt:md5,7385b870b955dae2c2ab78a70cf05cce"
                ]
            ]
        ],
        "timestamp": "2023-11-23T13:56:22.701059059"
    }
}
